### PR TITLE
Add interned string support to the JS<->WASM bindings layer

### PR DESCRIPTION
**Pull request by kg**
This PR introduces interned string support to the bindings layer along with some related changes:
* The JS bindings layer maintains a pair of string intern tables, one mapping JS string -> managed pointer and the other mapping managed pointer -> JS string.
* Any interned managed string is automatically added to the JS intern tables when it crosses the bindings boundary, to avoid duplicate allocations + reduce time spent copying/decoding text.
* Signatures gain a new type token ```S``` which specifies that the argument is an interned string. If the string passed is not already interned, the JS bindings will automatically intern it on both sides.
* The bindings layer logic for creating a new managed string will automatically look it up in the local intern table (unless it is large) and use the existing managed string if an interned string is found.
* mono_wasm_get_obj_type and similar APIs now return a distinct type ID for interned strings. The logic for this is awkward but doing it in the driver is *much* faster than trying to do it from JS.
* Partial support for converting JS ```Symbol``` instances to interned managed strings in the bindings layer. I didn't finish this yet because v8 shell's Symbol support is broken, but it's a good way to achieve interning because the JS runtime exposes interning for symbols.

I added some test coverage but I have some ideas for more tests to add, and I want to complete the symbol support.

### CODE REVIEW

**Comment by vargaz** [comment]
The mono parts look ok to me.

File: src/mono/wasm/runtime/binding_support.js [hunk 1]
```diff
@@ -1576,7 +1679,7 @@ var BindingSupportLib = {
 			return BINDING.js_string_to_mono_string ("Invalid JS object handle '" + js_handle + "'");
 		}
 
-		var js_name = BINDING.conv_string (method_name);
+		var js_name = BINDING.conv_string (method_name, false);
```
Comment (lewing) [inline src/mono/wasm/runtime/binding_support.js:1682]: Method name is extremely likely to be interned here.  We should either check here or add a flag to Interop/Runtime::InvokeJSWithArgs

Comment (kg) [reply src/mono/wasm/runtime/binding_support.js:1682]: I'm open to just interning it, this was an attempt to preserve behavior. I can't think of many cases where you would not want to intern the name of the method you're calling on a js object

Comment (lewing) [reply src/mono/wasm/runtime/binding_support.js:1682]: yeah, the paticular case I want to make sure we handle is stuff like
```csharp
        public int BytesPerElement => (int)GetObjectProperty("BYTES_PER_ELEMENT");
        public string Name => (string)GetObjectProperty("name");
        public int ByteLength => (int)GetObjectProperty("byteLength");
        public ArrayBuffer Buffer => (ArrayBuffer)GetObjectProperty("buffer");

        public void Fill(U value) => Invoke("fill", value);
        public void Fill(U value, int start) => Invoke("fill", value, start);
        public void Fill(U value, int start, int end) => Invoke("fill", value, start, end);

        public void Set(Array array) => Invoke("set", array);
        public void Set(Array array, int offset) => Invoke("set", array, offset);
        public void Set(ITypedArray typedArray) => Invoke("set", typedArray);
        public void Set(ITypedArray typedArray, int offset) => Invoke("set", typedArray, offset);

        public T Slice() => (T)Invoke("slice");
        public T Slice(int begin) => (T)Invoke("slice", begin);
        public T Slice(int begin, int end) => (T)Invoke("slice", begin, end);

        public T SubArray() => (T)Invoke("subarray");
        public T SubArray(int begin) => (T)Invoke("subarray", begin);
        public T SubArray(int begin, int end) => (T)Invoke("subarray", begin, end);
```
In the managed bindings

Comment (kg) [reply src/mono/wasm/runtime/binding_support.js:1682]: If those are passed as mono_obj (managed ptr) it should be easy to make that work, hopefully they're not being passed as C strings

Comment (lewing) [reply src/mono/wasm/runtime/binding_support.js:1682]: They are, and they go through this and the other paths (properties, global object, new, etc..) my first thought is something like https://gist.github.com/lewing/0cc29d338d8cc1f2239f31bb1f36360b which works but doesn't exactly make the code cleaner.

**Review by lewing — COMMENTED** [review]
This is a good start but I'd like to make sure we catch things from the managed side as mentioned in the comments.

**Review by lewing — APPROVED** [review]
Looks good other than some nits on the c formatting.  I will follow up with a PR for the other change I'd like to see.

**Pull request merged**